keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,415 @@
+Let's say we have a 3-zone cluster, and there is no pod in any node yet. Here
+comes a pod, and it wants to be scheduled to a zone which has pods with label
+`foo`. Obviously, there is no qualified node. However, we don't stop here;
+instead, we proceed to check if the incoming pod matches itself on its labels.
+If it does, we would think any node is a fit.
+
+This is actually an existing implication in PodAffinity algorithm. I just want to
+put here again to avoid confusion. And **below examples are all based the
+assumption that incoming pod matches itself on its labels**.
+
+- matching number and min matching number
+
+"matching" number is the number of pods matched on topology domain (defined by
+the global topologyKey). Suppose we have a 3-zone cluster, and there are 3 pods
+in zone1, 2 pods in zone2, 1 pod in zone3. And all pods carry label `foo`:
+
+```
++----------------------------+----------------------------+--------+
+|            zone1           |            zone2           |  zone3 |
++----------------------------+----------------------------+--------+
+| node1a |  node1b  | node1c |  node2a  | node2b | node2c | node3a |
++--------+----------+--------+----------+--------+--------+--------+
+|   pod  | pod, pod |        | pod, pod |        |        |   pod  |
++--------+----------+--------+----------+--------+--------+--------+
+```
+
+Now let's say there comes a pod, it wants to be placed along with pods which
+carries label `foo` in zones.
+
+If global topologyKey is "zone" and maxSkew is "1", then incoming pod can only
+be put into zone3 because for zone1, it violates `matching num (3) - min
+matching num (1) < maxSkey (1). Zone2 violate the formula the same way.
+
+If global topologyKey is "node" and masSkew is "1", things are slightly
+different. Min matching num becomes 0 now, and hence only node1c, node2a and
+node2c are qualified candidates.
+
+- what if a topology domain is infeasible
+
+Suppose we have pods distribution in a 3-zone cluster as 3/3/0, and all pods
+have label `foo`:
+
+```
++-------------+-------------+--------------------+
+|    zone1    |    zone2    | zone3 (infeasible) |
++-------------+-------------+--------------------+
+| pod,pod,pod | pod,pod,pod |                    |
++-------------+-------------+--------------------+
+```
+
+And we have an incoming pod which wants to be scheduled with pods which carry
+label `foo` in zones. And suppose all nodes in zone3 are infeasible, e.g. due to
+taints or lack of resources. In this case, we treat the `min matching num` as 3
+instead of 0, otherwise incoming pod failed to be scheduled with the presence of
+available resources in zone1 and zone2 - which is not what we want.
+
+- when formula check is enforced
+
+We only enforce the formula check upon new pod scheduling. In other words, if
+pods become imbalanced (due to explicit taints, lack of resources, or node
+lost), we don't do proactive re-scheduling. Our goal is to not make things
+worse.
+
+### Algorithm
```

Comment:
The name of this section is misleading. This is not the algorithm for the proposed design. `Use-cases` or `How user stories are addressed` or something like that would be better names.  In this section you are trying to say how this design addresses the user stories.

BTW, the user stories are a bit confusing too. They are mainly created for Design A. These user stories combine other affinities with even spreading. I think even spreading should only care about nodes of the cluster and have a mechanism to match pods in the specified node groups. The matching mechanism is a pod selector. That's it. We should not go any further than this for this feature. Users can combine this feature with other affinities in their pod spec if needed.

Reply:
Yes, as nobody objects to going towards Design B, I will keep Design B in the main body, and move Design A into "alternatives" section.

Reply:
Done.